pkg/controller/podgc/gc_controller.go

```diff
@@ -185,6 +186,114 @@ func (gcc *PodGCController) gcUnscheduledTerminating(pods []*v1.Pod) {
+}
+
+func NewDeletedNodesWatcher(kubeClient clientset.Interface) *deletedNodesWatcher {
+	return &deletedNodesWatcher{
+		recentlyDeletedNodes: make(map[string]time.Time),
+		existingNodes:        sets.NewString(),
+		kubeClient:           kubeClient,
+	}
+}
+
+func (dnw *deletedNodesWatcher) FilterDeletedNames(names sets.String) sets.String {
+	dnw.purgeIfNeeded()
+	deletedNames := sets.NewString()
+	unknownNames := sets.NewString()
+	// Filter against watch data
+	for name := range names {
+		if name == "" {
+			continue
+		}
+		if _, found := dnw.recentlyDeletedNodes[name]; found {
+			deletedNames.Insert(name)
+			continue
+		}
+		if _, found := dnw.existingNodes[name]; found {
+			continue
+		}
+		unknownNames.Insert(name)
+	}
+	if len(unknownNames) > 0 {
+		// If watch data is not enough, update existing nodes list
+		if err := dnw.updateExistingNodes(); err == nil {
+			// Filter against existing nodes, add to deleted if missing
+			var newDeletedNames []string
+			for name := range unknownNames {
+				if _, found := dnw.existingNodes[name]; found {
+					continue
+				}
+				deletedNames.Insert(name)
+				newDeletedNames = append(newDeletedNames, name)
+				dnw.markNodeDeleted(name)
+			}
+			if len(newDeletedNames) > 0 {
+				klog.Infof("Found new deleted nodes: %v", newDeletedNames)
+			}
+		}
+	}
+	return deletedNames
+}
+
+func (dnw *deletedNodesWatcher) updateExistingNodes() error {
+	klog.Infof("Updating existing nodes")
+	nodesList, err := dnw.kubeClient.CoreV1().Nodes().List(metav1.ListOptions{})
+	if err != nil {
+		klog.Errorf("Error while listing all Nodes: %v", err)
+		return err
+	}
+	for i := range nodesList.Items {
+		dnw.existingNodes.Insert(nodesList.Items[i].Name)
+	}
+	return nil
+}
+
+func (dnw *deletedNodesWatcher) purgeIfNeeded() {
+	if !time.Now().Add(-purgePeriod).After(dnw.lastPurge) {
+		return
+	}
+	removedCount := 0
+	cutOffTime := time.Now().Add(-purgeAfter)
+	for name, created := range dnw.recentlyDeletedNodes {
+		if cutOffTime.After(created) {
+			delete(dnw.recentlyDeletedNodes, name)
+			removedCount += 1
+		}
+	}
+	klog.Infof("Removed %v old deleted nodes from cache", removedCount)
+	dnw.lastPurge = time.Now()
+}
+
+func (dnw *deletedNodesWatcher) RegisterEventHandlers(nodeInformer coreinformers.NodeInformer) {
+	// Use shared informer to listen to add/update of nodes.
+	// Only events after registering will be visible.
+	nodeInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    dnw.addNode,
+		DeleteFunc: dnw.deleteNode,
+	})
+}
+
+func (dnw *deletedNodesWatcher) addNode(obj interface{}) {
+	node := obj.(*v1.Node)
+	dnw.existingNodes.Insert(node.Name)
+	delete(dnw.recentlyDeletedNodes, node.Name)
+}
+
+func (dnw *deletedNodesWatcher) deleteNode(obj interface{}) {
+	node := obj.(*v1.Node)
+	dnw.existingNodes.Delete(node.Name)
+	dnw.markNodeDeleted(node.Name)
+}
+
+func (dnw *deletedNodesWatcher) markNodeDeleted(name string) {
+	dnw.recentlyDeletedNodes[name] = time.Now()
```

Comment:
Should the recentlyDeletedNodes and existingNodes be protected by locks?

Reply:
Right, added guards